File: JavaScript/6js 1/findEvenIndex.js
/*
|--------------------------------------------------------------------------
| Equal Sides Of An Array
|--------------------------------------------------------------------------
|
*/

function findEvenIndex(arr) {
   for (let i = 0; i < arr.length; i++) {
      if (
         arr.slice(0, i).reduce((x, y) => x + y, 0) === arr.slice(i + 1).reduce((x, y) => x + y, 0)
      ) {
         return i;
      }
   }
   return -1;
}


function findEvenIndex(arr) {
   return arr.findIndex(
      (e, i, a) =>
         a.slice(0, i).reduce((p, c) => p + c, 0) == a.slice(i + 1).reduce((p, c) => p + c, 0)
   );
}


function findEvenIndex(arr) {
   let left = 0;
   let right = arr.reduce((s, n) => s + n, 0);
   for (let i = 0; i < arr.length; i++) {
      right -= arr[i];
      if (left === right) return i;
      left += arr[i];
   }
   return -1;
}


console.log(findEvenIndex([1, 2, 3, 4, 3, 2, 1]), 3, "The array was: [1,2,3,4,3,2,1] \n");
console.log(findEvenIndex([1, 100, 50, -51, 1, 1]), 1, "The array was: [1,100,50,-51,1,1] \n");
console.log(findEvenIndex([1, 2, 3, 4, 5, 6]), -1, "The array was: [1,2,3,4,5,6] \n");

File: Python/6py 1/find_even_index.py
"""
|--------------------------------------------------------------------------
| Equal Sides Of An Array
|--------------------------------------------------------------------------
|
"""

def find_even_index(arr):
   for i in range(len(arr)):
      if sum(arr[:i]) == sum(arr[i+1:]): return i
   return -1


find_even_index = lambda arr: next((i for i, __ in enumerate(arr) if sum(arr[:i]) == sum(arr[i+1:])), -1)


def find_even_index(arr):
   left, right = 0, sum(arr)
   for i, e in enumerate(arr):
      right -= e
      if left == right: return i
      left += e
   return -1


print(find_even_index([1,2,3,4,3,2,1]),3)
print(find_even_index([1,100,50,-51,1,1]),1,)
print(find_even_index([1,2,3,4,5,6]),-1)
print(find_even_index([20,10,30,10,10,15,35]),3)
print(find_even_index([20,10,-80,10,10,15,35]),0)